src/libraries/System.Private.CoreLib/src/System/Threading/Thread.Unix.cs

```diff
@@ -38,11 +38,77 @@ internal static void PollWasiEventLoopUntilResolvedVoid(Task mainTask)
         // the closest analog to Sleep(0) on Unix is sched_yield
         internal static void UninterruptibleSleep0() => Thread.Yield();
 
-#if !CORECLR
         private static void SleepInternal(int millisecondsTimeout) => WaitSubsystem.Sleep(millisecondsTimeout);
+
+#if !MONO
+        private bool JoinInternal(int millisecondsTimeout)
+        {
+            // This method assumes the thread has been started
+            Debug.Assert((ThreadState & ThreadState.Unstarted) == 0 || (millisecondsTimeout == 0));
+            SafeWaitHandle waitHandle = GetJoinHandle();
+
+            // If an OS thread is terminated and its Thread object is resurrected, waitHandle may be finalized and closed
+            if (waitHandle.IsClosed)
+            {
+                return true;
+            }
+
+            // If the current thread is the finalizer thread,
+            // we need to cleanup detached threads on another thread.
+            // We can't cleanup on the finalizer thread itself if we're on the finalizer thread,
+            // otherwise we could get into a deadlock waiting on the cleanup of a detached thread from a finalizer.
+            if (CurrentThreadIsFinalizerThread())
+            {
+                EnsureDetachedThreadCleanupThreadExists();
+            }
+
+            // Prevent race condition with the finalizer
+            try
+            {
+                waitHandle.DangerousAddRef();
+            }
+            catch (ObjectDisposedException)
+            {
+                return true;
+            }
+
+            try
+            {
+                return WaitSubsystem.Wait(waitHandle.DangerousGetHandle(), millisecondsTimeout, interruptible: false) == WaitHandle.WaitSuccess;
+            }
+            finally
+            {
+                waitHandle.DangerousRelease();
+            }
+        }
+
+        private void SetJoinHandle()
+        {
+            SafeWaitHandle waitHandle = GetJoinHandle();
+            Debug.Assert(!waitHandle.IsClosed);
+
+            waitHandle.DangerousAddRef();
+            try
+            {
+                WaitSubsystem.SetEvent(waitHandle.DangerousGetHandle());
+            }
+            finally
+            {
+                waitHandle.DangerousRelease();
+            }
+        }
 #endif
 
         // sched_getcpu doesn't exist on all platforms. On those it doesn't exist on, the shim returns -1
         internal static int GetCurrentProcessorNumber() => Interop.Sys.SchedGetCpu();
+
+        internal static void EnsureDetachedThreadCleanupThreadExists()
+        {
+            // We should only need to use a separate cleanup thread if we're on the finalizer thread.
```

Comment:
Do we still have the deadlock problem if the user code is waiting on a finalizer thread using hand-rolled synchronization (not one of ours) and the action that it is waiting for is blocked on observing an abandoned mutex?

Reply:
If the only time the user blocks the finalizer thread is by being blocked on another thread that blocks on detecting another thread dying, we would hit the same problem, you're right.

I'll remove the extra thread and add a comment in the finalizer thread logic with a warning about what the problem is so if someone ever reports this we can figure out the problem.